test/Microsoft.AspNet.Mvc.ModelBinding.Test/BodyBindingAndValidationIntegrationTest.cs

```diff
@@ -0,0 +1,186 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.ComponentModel.DataAnnotations;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http;
+using Microsoft.Framework.DependencyInjection;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Test
+{
+    public class BodyValidationIntegrationTests
+    {
+        private class Person
+        {
+            [FromBody]
+            [Required]
+            public Address Address { get; set; }
+        }
+
+        private class Address
+        {
+            public string Street { get; set; }
+        }
+
+        [Fact]
+        public async Task BodyBoundOnProperty_RequiredOnProperty_AddsModelStateError()
+        {
+            // Arrange
+            var argumentBinder = ModelBindingTestHelper.GetArgumentBinder();
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "Parameter1",
+                BindingInfo = new BindingInfo()
+                {
+                    BinderModelName = "CustomParameter",
+                },
+                ParameterType = typeof(Person)
+            };
+
+            var operationContext = ModelBindingTestHelper.GetOperationBindingContext();
+            var httpContext = operationContext.HttpContext;
+            ConfigureHttpRequest(httpContext.Request, string.Empty);
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            var model = await argumentBinder.BindModelAsync(parameter, modelState, operationContext);
```

Comment:
I think the checklist here for these tests to verify (unless N/A):
1). Model Binding Result (null/not-null, success/failure)
2). Model Values
3). ModelState.IsValue
4). Individual model keys

The point of these tests is to lock down these contracts in depth, so verifying just the things we think are significant for the test is a trap.